ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -117,27 +118,92 @@ where
         <λ as Add>::Output: Add<U1>,
         <<λ as Add>::Output as Add<U1>>::Output: ArrayLength,
         <λ as Add<U1>>::Output: ArrayLength,
+        I: IntoIterator<Item = F>,
+        J: IntoIterator<Item = F>,
+        I::IntoIter: ExactSizeIterator,
+        J::IntoIter: ExactSizeIterator,
     {
-        assert_eq!(self.u.len(), λ::USIZE); // We should pad with zeroes eventually
+        let mut u = u.into_iter();
+        let mut v = v.into_iter();
+
+        assert_eq!(u.len(), λ::USIZE); // We should pad with zeroes eventually
+        assert_eq!(v.len(), λ::USIZE); // We should pad with zeroes eventually
 
         // We need a table of size `λ + 1` since we add a random point at x=0
         let denominator = CanonicalLagrangeDenominator::<F, Sum<λ, U1>>::new();
         let lagrange_table = LagrangeTable::<F, Sum<λ, U1>, λ>::from(denominator);
 
-        let mut p = vec![p_0];
-        p.extend_from_slice(&self.u);
-        let mut q = vec![q_0];
-        q.extend_from_slice(&self.v);
+        let mut p = GenericArray::<F, Sum<λ, U1>>::generate(|_| F::ZERO);
+        let mut q = GenericArray::<F, Sum<λ, U1>>::generate(|_| F::ZERO);
+        let mut proof: GenericArray<F, TwoNPlusOne<λ>> = GenericArray::generate(|_| F::ZERO);
+        p[0] = p_0;
+        q[0] = q_0;
+        proof[0] = p_0 * q_0;
+
+        for i in 0..λ::USIZE {
+            let x = u.next().unwrap_or(F::ZERO);
+            let y = v.next().unwrap_or(F::ZERO);
+            p[i + 1] = x;
+            q[i + 1] = y;
+            proof[i + 1] += x * y;
+        }
         let p_extrapolated = lagrange_table.eval(&p);
         let q_extrapolated = lagrange_table.eval(&q);
 
-        ZeroKnowledgeProof::new(
-            zip(
-                p.into_iter().chain(p_extrapolated),
-                q.into_iter().chain(q_extrapolated),
-            )
-            .map(|(a, b)| a * b),
-        )
+        for (i, (x, y)) in zip(p_extrapolated.into_iter(), q_extrapolated.into_iter()).enumerate() {
+            proof[λ::USIZE + 1 + i] += x * y;
+        }
+
+        ZeroKnowledgeProof::new(proof)
+    }
+
+    pub fn gen_challenge_and_recurse<λ: ArrayLength, I, J>(
```

Comment:
I think this is the only function that actually uses the hashing right? It includes computing the challenge and compress the u and v values right?

Reply:
That's right.